File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> freq;
        for (int &num : nums) {
            freq[num]++;
        }

        vector<int> res;

        priority_queue<pair<int, int>> pq;
        for (auto &[key, value] : freq) {
            pq.push({value, key});
        }

        while (k--) {
            res.push_back(pq.top().second);
            pq.pop();
        }
        return res;

    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pref = [0] * n
        suff = [0] * n
        pref[0] = suff[n - 1] = 1

        for i in range(1, len(nums)):
            pref[i] = nums[i - 1] * pref[i - 1] 
        for i in range(len(nums) - 2, -1, -1):
            suff[i] = nums[i + 1] * suff[i + 1]
        res = [0] * n
        for i in range(n):
            res[i] = pref[i] * suff[i] 
        return res